x-pack/elastic-agent/pkg/agent/application/upgrade/upgrade.go

```diff
@@ -147,32 +151,35 @@ func (u *Upgrader) Upgrade(ctx context.Context, a Action, reexecNow bool) (err e
 			u.ackAction(ctx, action)
 		}
 		u.log.Warn("upgrading to same version")
-		return nil
+		return nil, nil
 	}
 
 	if err := copyActionStore(newHash); err != nil {
-		return errors.New(err, "failed to copy action store")
+		return nil, errors.New(err, "failed to copy action store")
 	}
 
 	if err := ChangeSymlink(ctx, newHash); err != nil {
 		rollbackInstall(ctx, newHash)
-		return err
+		return nil, err
 	}
 
 	if err := u.markUpgrade(ctx, newHash, a); err != nil {
 		rollbackInstall(ctx, newHash)
-		return err
+		return nil, err
 	}
 
 	if err := InvokeWatcher(u.log); err != nil {
 		rollbackInstall(ctx, newHash)
-		return errors.New("failed to invoke rollback watcher", err)
+		return nil, errors.New("failed to invoke rollback watcher", err)
 	}
 
+	cb := shutdownCallback(u.log, release.Version(), a.Version(), release.TrimCommit(newHash))
 	if reexecNow {
-		u.reexec.ReExec()
+		u.reexec.ReExec(cb)
+		return nil, nil
```

Comment:
because we need to make sure callback is called on shutdown, if reexec now is set we call reexec with a callback
if it's not set we dont know when reexec will be called. so we return callback to make caller responsible to properly closing agent